schnorr_fun/src/frost/shared_key.rs

```diff
@@ -0,0 +1,410 @@
+    #[must_use]
+    pub fn homomorphic_add(
+        mut self,
+        tweak: Scalar<impl Secrecy, impl ZeroChoice>,
+    ) -> SharedKey<Normal, Zero> {
+        self.point_polynomial[0] = g!(self.point_polynomial[0] + tweak * G).normalize();
+        SharedKey::from_inner(self.point_polynomial)
+    }
+
+    /// Negates the polynomial
+    #[must_use]
+    pub fn homomorphic_negate(mut self) -> SharedKey<Normal, Z> {
+        poly::point::negate(&mut self.point_polynomial);
+        SharedKey::from_inner(self.point_polynomial)
+    }
+
+    /// Multiplies the shared key by a scalar.
+    ///
+    /// In order for a [`PairedSecretShare`] to be valid against the new key they will have to apply
+    /// [the same operation](super::PairedSecretShare::homomorphic_mul).
+    ///
+    /// [`PairedSecretShare`]: super::PairedSecretShare
+    #[must_use]
+    pub fn homomorphic_mul(mut self, tweak: Scalar<impl Secrecy>) -> SharedKey<Normal, Z> {
+        for coeff in &mut self.point_polynomial {
+            *coeff = g!(tweak * coeff.deref()).normalize();
+        }
+        SharedKey::from_inner(self.point_polynomial)
+    }
+
+    /// Create a shared key from a subset of verification shares.
+    ///
+    /// If all the verification shares are correct and you have at least a threshold of them then
+    /// you'll get the right answer. If you put in a wrong share you won't get the right answer!
+    ///
+    /// ## Security
+    ///
+    /// ⚠ You can't just take any random points you want and pass them in here and hope it's secure.
+    /// They need to be from a securely generated key.
+    pub fn from_verification_shares(
+        shares: &[(PartyIndex, Point<impl PointType, Public, impl ZeroChoice>)],
+    ) -> SharedKey<Normal, Z> {
+        let poly = poly::point::interpolate(shares);
+        let poly = poly::point::normalize(poly);
+        SharedKey::from_inner(poly.collect())
+    }
+
+    /// The public key that has been shared.
+    ///
+    /// This is using *public key* in a rather loose sense. Unless it's a `SharedKey<EvenY>` then it
+    /// won't be usable as an actual Schnorr [BIP340] public key.
+    ///
+    /// [BIP340]: https://bips.xyz/340
+    pub fn public_key(&self) -> Point<T, Public, Z> {
+        // SAFETY: we hold the first coefficient to match the type parameters always
+        let public_key = Z::cast_point(self.point_polynomial[0]).expect("invariant");
+        T::cast_point(public_key).expect("invariant")
+    }
+
+    /// Encodes a `SharedKey` as the compressed encoding of each underlying polynomial coefficient
+    ///
+    /// i.e. call [`Point::to_bytes`] on each coefficent starting with the constant term. Note that
+    /// even if it's a `SharedKey<EvenY>` the first coefficient (A.K.A the public key) will still be
+    /// encoded as 33 bytes.
+    ///
+    /// ⚠ Unlike other secp256kfun things this doesn't exactly match the serde/bincode
+    /// implemenations which will length prefix the list of points.
```

Comment:
I am curious as to why? Or is this just a note that serde/bincode prefix the length so you'll have to keep that in mind if you mix & match?